### DISCUSSION TITLE
Locking down the GraphQL endpoint

### DISCUSSION THREAD

**Discussion by chirino** [guacsec/trustify#642] [General]
I noticed that the project has a GraphQL endpoint.

I'm a big fan of GraphQL for locked down aka (white listed queries). I'm not a fan of it for ad-hoc queries because It's typically easy to construct queries that can DOS your app.  So do we have any strategies to mitigate this issue?

**Comment by i386x** [comment]
>  that can DOS your app.

Hijacking the discussion. Recently I found [this commit](https://github.com/trustification/trustify/commit/11b0edea39573c9ff26d59ec3eca2d6658020354#diff-39ae66f11a25c215898e779828aa61a8f3325989820a7ff0229acc2e8704236b) and I am a bit afraid if someone can upload hand-crafted SBOMs and then initiate delete operation with an intention to DOS application on the worst gc performance scenario? Based on theory, a (directed) graph with N nodes can have at most N\*N edges so traversing such a graph-like data structure, like the one stored in Postgres database, may also require N\*N operations. On the other hand the same number of operations can be probably needed when storing a sinister SBOMs to the database so my concerns are probably just a false alarm. What do you think?

**Comment by helio-frota** [comment]
Can we close this?

**Reply by ctron** [reply]
I think it's still a valid one.

**Comment by helio-frota** [comment]
I guess they don't have rate-limiter implemented yet
https://github.com/async-graphql/async-graphql/issues/1618
https://docs.rs/async-graphql/latest/async_graphql/?search=limit

**Comment by helio-frota** [comment]
Created this issue https://github.com/trustification/trustify/issues/1228
(not saying that I'm going to do this -- right now)

**Reply by helio-frota** [reply]
> So do we have any strategies to mitigate this issue?

The issue title includes "investigate." I don't want to imply that rate-limiting is the ideal solution, but we need alternatives for mitigation :+1:

**Reply by helio-frota** [reply]
as a 'ping' I closed the issue and marked as invalid :+1:
